DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,242 @@
+```D
+enum runMyTest;
+
+@runMyTest unittest {
+
+}
+```
+Since most of these declarations are at or near module scope, we can simply iterate over *all* the `unittest`s looking for one's matching our UDA. Although not technically required, it is still common to find these tests in a pre-compilation step - or by manually setting the test runner to work using a `mixin` statement.
+
+This pattern is fine for `unittest`s: When used, there are usually many of them, and when there are not many the potential for a performance hit is mitigated by the compiler providing a list of `unittest`s within a given scope.
+
+### The behaviour to be enabled by this DIP.
+There are, however, situations where this pattern is less than ideal. Some UDAs may be used to declare things which are both sparse and not `unittest`s. 
+
+If we setup a mechanism to get these UDA-ed symbols - assuming we know where they are i.e. a recursive search is currently not possible but scopes containing UDAs can be declared using a different UDA - we still have to iterate over *every* declaration in a given scope looking for those with our desired set of attributes. For a sparse declaration - that is, if we are looking for a few declarations out of a several thousand line file - this is not an efficient way of doing things.
+
+For example, this DIP is motivated specifically by a desire to be able to declare benchmarks in the following manner (without needing the user to `mixin` anything):
+
+```D
+@Benchmark!(SomeInformationAboutTheBenchmark)
+float dotp(float[3] vec1, float[3] vec2) pure
+{
+    return vec1[0] * vec2[0] + vec1[1] * vec2[1] + vec1[2] * vec2[2] 
+}
+//The library can then do what it wants with dotp, be that schedule it to run or add it to a table at compile time
+```
+
+This DIP proposes a simple solution to this problem: Let the UDA see sideways, that is, let it see what it is attached to. Rather than being some paradigm shift in how UDAs are used, this should simply be viewed as having the same effect of adding syntactic sugar for the following construct:
+```D
+    module home;
+    enum MyUDA;
+    template HandleTheUDA(alias handleThis) {/* impl */}
+    @MyUDA void widget() {}
+    //Ugly
+    mixin HandleTheUDA!handleThis;
+    //Slow, we don't want to search the entire module just for one UDA
+    mixin HandleTheUDA!home;
+```
+i.e. With this DIP, MyUDA can be declared as a template that can perform the `mixin` itself. This does not add any new dragons or side effects beyond what the original construct could do.
+## Prior Work
+### A pattern that already does the job
+The following pattern can be used within a UDA to find what said UDA is attached to.
+```D
+//Module is needed for obvious reasons, the line parameter makes it unique for most uses
+template FinderUDA(string name, string m = __MODULE__, int l = __LINE__)
+{
+    import std.format;
+	import std.traits;
+    enum FinderUDA;
+    mixin(format!"alias mod = %s;"(m));
```

Comment:
Better use `mixin("alias mod = ", m);` or at least `mixin("alias mod = " ~ m);`. Using `format` is unnecessary and encourages that use.